flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/HITSAlgorithm.java

```diff
@@ -0,0 +1,194 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.graph.library;
+
+import org.apache.flink.api.common.aggregators.DoubleSumAggregator;
+import org.apache.flink.api.java.DataSet;
+
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.GraphAlgorithm;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.graph.spargel.MessageIterator;
+import org.apache.flink.graph.spargel.MessagingFunction;
+import org.apache.flink.graph.spargel.ScatterGatherConfiguration;
+import org.apache.flink.graph.spargel.VertexUpdateFunction;
+import org.apache.flink.types.DoubleValue;
+
+/**
+ * This is an implementation of HITS algorithm, using a scatter-gather iteration.
+ * The user can define the maximum number of iterations. HITS algorithm is determined by two parameters,
+ * hubs and authorities. A good hub represented a page that pointed to many other pages, and a good authority
+ * represented a page that was linked by many different hubs. The implementation assumes that the two value on
+ * every vertex are the same at the beginning.
+ * <p>
+ * If the number of vertices of the input graph is known, it should be provided as a parameter
+ * to speed up computation. Otherwise, the algorithm will first execute a job to count the vertices.
+ */
+public class HITSAlgorithm<K> implements GraphAlgorithm<K, Double, Double, DataSet<Vertex<K, Double>>> {
+
+	public static enum HITSParameter {
+		HUB,
+		AUTHORITY
+	}
+
+	private int maxIterations;
+	private long numberOfVertices;
+
+	/**
+	 * Creates an instance of HITS algorithm.
+	 * If the number of vertices of the input graph is known,
+	 * use the {@link HITSAlgorithm#HITSAlgorithm(int, long, HITSParameter)} constructor instead.
+	 *
+	 * @param maxIterations the maximum number of iterations
+	 * @param hitsParameter the type of final web pages users want to get by this algorithm
+	 */
+	public HITSAlgorithm(int maxIterations, HITSParameter hitsParameter) {
+		if (hitsParameter == HITSParameter.AUTHORITY) {
+			this.maxIterations = maxIterations * 2;
+		} else {
+			this.maxIterations = maxIterations * 2 + 1;
+		}
+	}
+
+	/**
+	 * Creates an instance of HITS algorithm.
+	 * If the number of vertices of the input graph is unknown,
+	 * use the {@link HITSAlgorithm#HITSAlgorithm(int, HITSParameter)} constructor instead.
+	 *
+	 * @param maxIterations the maximum number of iterations
+	 * @param hitsParameter the type of final web pages users want to get by this algorithm
+	 */
+	public HITSAlgorithm(int maxIterations, long numberOfVertices, HITSParameter hitsParameter) {
+		if (hitsParameter == HITSParameter.AUTHORITY) {
+			this.maxIterations = maxIterations * 2;
+		} else {
+			this.maxIterations = maxIterations * 2 + 1;
+		}
+		this.numberOfVertices = numberOfVertices;
```

Comment:
Verify positive `numberOfVertices` with `Preconditions.checkArgument`. Same for `maxIterations` in other constructor.